### PR TITLE
Reachability notifications

### PR DESCRIPTION
**Pull request by err508**
## Description

 Adds reachability_announcement and handling to enhance UserPresence tracking. Specifically:
- toDevice message was rewritten to ReachabilityNotification
- new periodic Greenlet to send a ReachabilityNotification to all
known peers every 10 seconds
- _handle_reachability_announcement to update the UserPresence in
case it is incorrect.

## PR review check list

Quality check list that cannot be automatically verified.

- Safety
    - [x] The changes respect the necessary conditions for safety (https://raiden-network-specification.readthedocs.io/en/latest/smart_contracts.html#protocol-values-constraints)
-  Code quality
    - [ ] Error conditions are handled
    - [x] Exceptions are propagated to the correct parent greenlet
    - [ ] Exceptions are correctly classified as recoverable or unrecoverable
- Compatibility
    - [ ] State changes are forward compatible
    - [ ] Transport messages are backwards and forward compatible
- Commits
    - [x] Have good messages
    - [x] Squashed unecessary commits
- If it's a bug fix:
    - Regression test for the bug
        - [ ] Properly covers the bug
        - [x] If an integration test is used, it could not be written as a unit test
- Documentation
    - [ ] A new CLI flag was introduced, is there documentation that explains usage?
- Specs
    - [ ] If this is a protocol change, are the specs updated accordingly? If so, please link PR from the specs repo.
- Is it a user facing feature/bug fix?
    - [ ] Changelog entry has been added

### CODE REVIEW

File: raiden/tests/integration/network/transport/test_matrix_transport.py [hunk 1]
```diff
@@ -1080,7 +1079,11 @@ def test_reproduce_handle_invite_send_race_issue_3588(matrix_transports):
 
 @pytest.mark.parametrize("matrix_server_count", [1])
 @pytest.mark.parametrize("number_of_transports", [2])
-def test_send_to_device(matrix_transports):
+def test_send_reachability_notification(matrix_transports):
+    """
+    Basic test to test reachability notifications work
```
Comment (LefterisJP) [inline raiden/tests/integration/network/transport/test_matrix_transport.py]: nitpick: work -> works

File: raiden/messages/matrix.py [hunk 2]
```diff
@@ -3,24 +3,19 @@
...
 
 @dataclass(repr=False, eq=False)
-class ToDevice(SignedMessage):
-    """
-    Message, which can be directly sent to all devices of a node known by matrix,
-    no room required. Messages which are supposed to be sent via transport.sent_to_device must
-    subclass.
-    """
+class ReachabilityNotification(SignedMessage):
+    # FIXME Message needs a nonce/identifier to prevent replays/spoofing
```
Comment (LefterisJP) [inline raiden/messages/matrix.py]: Can you make an issue for this TODO?

File: raiden/messages/cmdid.py [hunk 3]
```diff
@@ -18,7 +18,7 @@ class CmdId(enum.Enum):
     REVEALSECRET = 11
     DELIVERED = 12
     LOCKEXPIRED = 13
-    TODEVICE = 14
+    REACHABILITYNOTIFICATION = 14
```
Comment (andrevmatos) [inline raiden/messages/cmdid.py]: Can this be split with `_`, for consistency?
```suggestion
    REACHABILITY_NOTIFICATION = 14
```

File: raiden/network/transport/matrix/transport.py [hunk 4]
```diff
@@ -1267,80 +1259,98 @@ def _leave_unused_rooms(self, _address_to_room_ids: Dict[AddressHex, List[_RoomI
...
+                )
+                self._raiden_service.sign(message)
+                for address in self._address_mgr.known_addresses:
+                    user_ids = self._address_mgr.get_userids_for_address(address)
+                    data = {
+                        user_id: {"*": MessageSerializer.serialize(message)}
+                        for user_id in user_ids
+                    }
+                    self._client.api.send_to_device("m.reachability_notification", data)
+                self._stop_event.wait(10)
```
Comment (andrevmatos) [inline raiden/network/transport/matrix/transport.py]: Should this be made a constant in constants.py or settings.py?

File: raiden/network/transport/matrix/transport.py [hunk 5]
```diff
@@ -1267,80 +1259,98 @@ def _leave_unused_rooms(self, _address_to_room_ids: Dict[AddressHex, List[_RoomI
...
-
-        return self._client.api.send_to_device("m.to_device_message", data)
+    def reachability_announcement(self, reachability: AddressReachability) -> None:
+        if self._raiden_service:
+            while not self._stop_event.ready():
+                self.log.debug("Announcing reachability to all known peers")
+                message = ReachabilityNotification(
+                    address_reachability=reachability.value, signature=EMPTY_SIGNATURE
+                )
+                self._raiden_service.sign(message)
```
Comment (andrevmatos) [inline raiden/network/transport/matrix/transport.py]: I have some strong concerns about this. Requiring a new signature to be made every 10s shouldn't be made without syncing with the Light Client team. This actually is very unfeasible for us, as all signatures are interactive (for now), and even after we implement subkey for non-interactive signing, the former case will still be supported and would be worse than it's already is with this change.

**Comment by rakanalh** [comment]
I agree with @andrevmatos. And as discussed before, this is not a solution that i would go for even if it was a temporary measure.

**Comment by err508** [comment]
I think we might have to do something similar in the future, as matrix presence is "best effort". We have a strong liveness assumption, in matrix there is no such thing.

**Pull request closed without merging**